source_handlers.go

```diff
@@ -331,3 +331,65 @@ func SourceCheckAvailability(c echo.Context) error {
 
 	return c.JSON(http.StatusAccepted, map[string]interface{}{})
 }
+
+// SourcePause pauses a source and all its dependant applications, by setting the former's and the latter's "paused_at"
+// columns to "now()".
+func SourcePause(c echo.Context) error {
+	applicationDao, err := getApplicationDao(c)
+	if err != nil {
+		return err
+	}
+
+	sourceId, err := strconv.ParseInt(c.Param("source_id"), 10, 64)
+	if err != nil {
+		return c.JSON(http.StatusBadRequest, util.ErrorDoc(err.Error(), "400"))
+	}
+
+	err = applicationDao.PauseAll(sourceId)
+	if err != nil {
+		return c.JSON(http.StatusBadRequest, util.ErrorDoc(err.Error(), "400"))
+	}
+
+	sourceDao, err := getSourceDao(c)
+	if err != nil {
+		return err
+	}
+
+	err = sourceDao.Pause(sourceId)
+	if err != nil {
+		return c.JSON(http.StatusBadRequest, util.ErrorDoc(err.Error(), "400"))
+	}
+
+	return c.JSON(http.StatusNoContent, nil)
```

Comment:
So during the `PauseAll` events one side effect is that we need to raise events. I foresee a service class that basically goes through and:
1. calls Pause on all the applications for the source
2. raises the event for each

old code here: https://github.com/RedHatInsights/sources-api/blob/7b8737277a1b7fcf0be2d58feb579272b66e7257/app/controllers/api/v3x1/sources_controller.rb#L21-L24


We could have a file in service that basically has 2 functions, `PauseSource` and `PauseApplication` where the PauseSource just calls PauseApplication for each source's applications.

Reply:
I thought about doing something reusable, but I ruled that possibility out due to the N+1 issue. I don't know if we will have many applications related to a source, but it can get out of hand pretty easily.

Instead I was thinking on maybe doing the following:

1. Update all the applications for a source.
2. Update the source itself.
3. Fetch the applications and loop through raising events for every single one of them.

I could wrap the first two operations in a transaction so that if something fails, no fake pause notifications are raised.

Same thing for the applications:

1. Update them.
2. Fetch them.
3. Raise events.

What do you think of this approach? This way we could be done with 2 to 3 queries max.

Reply:
Just to clarify, what I'm thinking of is the following thing:

```go
sourceId := 12345
source, applicationList, err := sourceDao.Pause(sourceId) // this pauses the related applications to the source and the source itself in a transaction

// raiseEvent("Source.Pause", source)
// for _, app := range applicationList { raiseEvent("Application.Pause", application) } 
```

And for the applications:

```go
appId := 54321
application, err := applicationDao.Pause(appId) // this pauses the application
// raiseEvent("Application.Pause", application)
```

Reply:
Yeah that sounds good to me. We can probably pull the actual `raiseEvent` piece into a helper function (it's still just a middleware right now).

Reply:
I'm resolving this since I've just implemented the changes.